### PR TITLE
Complete decoupling ANTLR AST from Painless AST

### PR DESCRIPTION
**Pull request by jdconrad**
This change removes the Reserved class used to track variables usages within the ANTLR grammar. That task is now performed by an existing pass "extractVariables" in the Painless AST. The Painless AST no longer has any dependencies on the ANTLR AST for state outside of the tree being built. This will simplify future refactoring and opens the possibility of alternate grammars.

### CODE REVIEW

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SFunction.java [hunk 1]
```diff
@@ -108,10 +86,16 @@ void storeSettings(CompilerSettings settings) {
         this.settings = settings;
     }
 
+    void extractVariables() {
```
Comment (rjernst) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SFunction.java]: couldn't the few callers of this just call this method themselves with an empty set?

**Review by rjernst — APPROVED** [review]
LGTM

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SFunction.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SFunction.java]: Yes, that makes more sense.  Will fix this.

**Comment by jdconrad** [comment]
@rjernst Thanks for the review!  Will commit as soon as CI passes.

**Comment by jdconrad** [comment]
@elasticmachine run elasticsearch-ci-/2

**Comment by jdconrad** [comment]
@elasticmachine elasticsearch-ci/2

**Comment by jdconrad** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by jdconrad** [comment]
@elasticmachine run elasticsearch-ci/1

**Pull request merged**